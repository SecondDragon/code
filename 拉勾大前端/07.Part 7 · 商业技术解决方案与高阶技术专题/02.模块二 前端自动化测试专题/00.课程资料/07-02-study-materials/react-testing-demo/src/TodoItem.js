import React, { useState, useRef } from 'react'
import classNames from 'classnames'

export default function TodoItem ({ todo, doneTodo, removeTodo, saveTodo }) {
  const [isEdit, setIsEdit] = useState(false)
  const inputEl = useRef(null)

  const saveEdit = e => {
    const value = e.target.value.trim()
    if (value.length) {
      saveTodo(todo.id, e.target.value)
    } else {
      removeTodo(todo.id)
    }
    setIsEdit(false)
  }

  const cancelEdit = e => {
    setIsEdit(false)
    e.target.value = todo.text
  }

  return (
    <li className={classNames({
      completed: todo.done,
      editing: isEdit
    })} data-testid="todo-item">
      {/* These are here just to show the structure of the list items */}
      {/* List items should get the class `editing` when editing and `completed` when marked as completed */}
      <div className="view">
        <input
          className="toggle"
          type="checkbox"
          checked={todo.done}
          data-testid="todo-done"
          onChange={e => doneTodo(todo.id, e.target.checked)}
        />
        <label
          data-testid="todo-text"
          onDoubleClick={() => {
            setIsEdit(true)
            setTimeout(() => {
              inputEl.current.focus()
            }, 0)
          }}
        >{todo.text}</label>
        <button
          className="destroy"
          data-testid="destroy"
          onClick={() => removeTodo(todo.id)}
        ></button>
      </div>
      <input
        ref={inputEl}
        className="edit"
        autoFocus
        defaultValue={todo.text}
        onBlur={saveEdit}
        data-testid="todo-edit"
        onKeyUp={e => {
          if (e.key === 'Enter') { // 保存编辑
            saveEdit(e)
          } else if (e.key === 'Escape') { // 取消编辑
            cancelEdit(e)
          }
        }}
      />
    </li>
  )
}
