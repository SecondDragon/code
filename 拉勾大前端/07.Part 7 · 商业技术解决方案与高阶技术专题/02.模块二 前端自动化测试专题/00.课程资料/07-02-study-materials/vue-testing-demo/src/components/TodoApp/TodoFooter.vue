<template>
  <footer class="footer">
    <!-- This should be `0 items left` by default -->
    <span class="todo-count"
      ><strong data-testid="done-todos-count">{{ doneTodosCount }}</strong> item
      left</span
    >
    <!-- Remove this if you don't implement routing -->
    <ul class="filters">
      <li>
        <!-- <a class="selected" href="#/">All</a> -->
        <router-link data-testid="link-all" exact to="/">All</router-link>
      </li>
      <li>
        <!-- <a href="#/active">Active</a> -->
        <router-link data-testid="link-active" to="/active">Active</router-link>
      </li>
      <li>
        <!-- <a href="#/completed">Completed</a> -->
        <router-link data-testid="link-completed" to="/completed">Completed</router-link>
      </li>
    </ul>
    <!-- Hidden if no completed items are left â†“ -->
    <button
      v-if="isClearCompletedShow"
      class="clear-completed"
      data-testid="clear-completed"
      @click="$emit('clear-completed')"
    >Clear completed</button>
  </footer>
</template>

<script>
export default {
  name: 'TodoFooter',
  props: {
    todos: {
      type: Array,
      required: true
    }
  },
  computed: {
    doneTodosCount () {
      return this.todos.filter(t => !t.done).length
    },
    isClearCompletedShow () {
      return this.todos.find(t => t.done)
    }
  }
}
</script>
