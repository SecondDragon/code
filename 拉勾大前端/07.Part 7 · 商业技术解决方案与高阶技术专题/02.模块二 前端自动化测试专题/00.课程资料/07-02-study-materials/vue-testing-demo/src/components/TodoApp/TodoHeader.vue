<template>
  <header class="header">
    <h1>todos</h1>
    <input
      class="new-todo"
      placeholder="What needs to be done?"
      autofocus
      data-testid="new-todo"
      @keyup.enter="handleNewTodo"
    />
  </header>
</template>

<script>
export default {
  name: 'TodoHeader',
  methods: {
    handleNewTodo (e) {
      const value = e.target.value.trim()
      if (!value.length) {
        return
      }
      this.$emit('new-todo', value)
      e.target.value = ''
    }
  }
}
</script>
