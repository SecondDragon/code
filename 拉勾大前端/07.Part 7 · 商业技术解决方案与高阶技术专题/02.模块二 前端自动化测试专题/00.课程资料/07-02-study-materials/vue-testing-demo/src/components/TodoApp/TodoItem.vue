<template>
  <li data-testid="todo-item" :class="{
    completed: todo.done,
    editing: isEditing
  }">
    <div class="view">
      <input v-model="todo.done" data-testid="todo-done" class="toggle" type="checkbox" />
      <label
        data-testid="todo-text"
        @dblclick="isEditing = true"
      >{{ todo.text }}</label>
      <button
        data-testid="delete"
        class="destroy"
        @click="$emit('delete-todo', todo.id)"
      ></button>
    </div>
    <input
      :value="todo.text"
      v-focus="isEditing"
      class="edit"
      data-testid="todo-edit"
      @blur="isEditing = false"
      @keyup.enter="handleEditTodo"
      @keyup.esc="handleCancelEdit"
    />
  </li>
</template>

<script>
export default {
  name: 'TodoItem',
  props: {
    todo: {
      type: Object,
      required: true
    }
  },
  directives: {
    focus (element, binding) {
      if (binding.value) {
        element.focus()
      }
    }
  },
  data () {
    return {
      isEditing: false
    }
  },
  methods: {
    handleEditTodo (e) {
      this.$emit('edit-todo', {
        id: this.todo.id,
        text: e.target.value
      })
      // 取消编辑状态
      this.isEditing = false
    },

    handleCancelEdit () {
      this.isEditing = false
    }
  }
}
</script>
