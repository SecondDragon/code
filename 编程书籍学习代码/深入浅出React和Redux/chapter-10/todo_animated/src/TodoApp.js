import React from 'react';
import {view as Todos} from './todos/';
import {view as Filter} from './filter/';

function TodoApp() {
  return (
    <div>
      <Todos />
      <Filter />
    </div>
  );
}

export default TodoApp;
