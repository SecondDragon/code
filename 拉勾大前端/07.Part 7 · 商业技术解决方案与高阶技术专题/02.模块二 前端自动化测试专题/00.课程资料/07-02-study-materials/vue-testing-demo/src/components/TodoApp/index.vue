<template>
  <section class="todoapp">
    <TodoHeader @new-todo="handleNewTodo"/>
    <!-- This section should be hidden by default and shown when there are todos -->
    <section class="main">
      <input
        v-model="toggleAll"
        data-testid="toggle-all"
        id="toggle-all"
        class="toggle-all"
        type="checkbox"
      />
      <label for="toggle-all">Mark all as complete</label>
      <ul class="todo-list">
        <!-- These are here just to show the structure of the list items -->
        <!-- List items should get the class `editing` when editing and `completed` when marked as completed -->
        <TodoItem
          v-for="todo in filterTodos"
          :key="todo.id"
          :todo="todo"
          @delete-todo="handleDelteTodo"
          @edit-todo="handleEditTodo"
        />
      </ul>
    </section>
    <!-- This footer should be hidden by default and shown when there are todos -->
    <TodoFooter
      :todos="todos"
      @clear-completed="handleClearCompleted"
    />
  </section>
</template>

<script>
import TodoHeader from './TodoHeader'
import TodoFooter from './TodoFooter'
import TodoItem from './TodoItem'

export default {
  name: 'TodoApp',
  components: {
    TodoHeader,
    TodoFooter,
    TodoItem
  },
  data () {
    return {
      todos: []
    }
  },
  computed: {
    toggleAll: {
      get () {
        // 设置 toggleALl 的选中状态
        return this.todos.length && this.todos.every(t => t.done)
      },
      set (checked) {
        this.todos.forEach(todo => {
          todo.done = checked
        })
      }
    },

    // 过滤数据
    filterTodos () {
      // 获取路由路径
      const path = this.$route.path
      // 根据路由路径过滤数据
      // / 所有的任务列表
      // /active 所有的未完成任务
      // /completed 所有的已完成任务
      let todos = this.todos
      switch (path) {
        case '/active':
          todos = this.todos.filter(t => !t.done)
          break
        case '/completed':
          todos = this.todos.filter(t => t.done)
          break
      }
      return todos
    }
  },
  methods: {
    handleNewTodo (text) {
      const lastTodo = this.todos[this.todos.length - 1]
      this.todos.push({
        id: lastTodo ? lastTodo.id + 1 : 1,
        text,
        done: false
      })
    },
    handleDelteTodo (todoId) {
      const index = this.todos.findIndex(t => t.id === todoId)
      if (index !== -1) {
        this.todos.splice(index, 1)
      }
    },
    handleEditTodo ({ id, text }) {
      const todo = this.todos.find(t => t.id === id)
      if (!todo) {
        return
      }
      if (!text.trim().length) {
        // 执行删除操作
        return this.handleDelteTodo(id)
      }
      // 执行修改操作
      todo.text = text
    },

    handleClearCompleted () {
      // 清除所有已完成的任务项
      for (let i = 0; i < this.todos.length; i++) {
        if (this.todos[i].done) {
          // 执行删除操作
          this.todos.splice(i, 1)
          i--
        }
      }
    }
  }
}
</script>
