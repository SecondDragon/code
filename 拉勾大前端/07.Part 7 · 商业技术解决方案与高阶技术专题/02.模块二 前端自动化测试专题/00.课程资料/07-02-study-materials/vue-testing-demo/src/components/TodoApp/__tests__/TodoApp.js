import { shallowMount } from '@vue/test-utils'
import TodoApp from '@/components/TodoApp/index.vue'
import TodoItem from '@/components/TodoApp/TodoItem.vue'
import Vue from 'vue'

describe('TodoApp.vue', () => {
  /** @type {import('@vue/test-utils').Wrapper} */
  let wrapper = null

  beforeEach(async () => {
    const $route = {
      path: '/'
    }
    wrapper = shallowMount(TodoApp, {
      mocks: {
        $route
      }
    })
    const todos = [
      { id: 1, text: 'eat', done: false },
      { id: 2, text: 'play', done: true },
      { id: 3, text: 'sleep', done: false }
    ]
    await wrapper.setData({
      todos
    })
  })

  test('New todo', async () => {
    const text = 'play'
    wrapper.vm.handleNewTodo(text)
    const todo = wrapper.vm.todos.find(t => t.text === text)
    expect(todo).toBeTruthy()
  })

  test('Todo List', async () => {
    expect(wrapper.findAllComponents(TodoItem).length).toBe(wrapper.vm.todos.length)
  })

  test('Delete Todo', async () => {
    await wrapper.vm.handleDelteTodo(1)
    expect(wrapper.vm.todos.length).toBe(2)
    expect(wrapper.findAllComponents(TodoItem).length).toBe(2)
  })

  test('Delete Todo', async () => {
    await wrapper.vm.handleDelteTodo(123)
    expect(wrapper.vm.todos.length).toBe(3)
    expect(wrapper.findAllComponents(TodoItem).length).toBe(3)
  })

  test('Edit Todo', async () => {
    const todo = { id: 2, text: 'abc' }
    await wrapper.vm.handleEditTodo(todo)
    expect(wrapper.vm.todos[1].text).toBe(todo.text)

    todo.text = ''
    await wrapper.vm.handleEditTodo(todo)
    expect(wrapper.vm.todos.find(t => t.id === todo.id)).toBeFalsy()
  })

  test('Toggle All', async () => {
    const toggleAll = wrapper.find('input[data-testid="toggle-all"]')
    await toggleAll.setChecked()
    // 断言所有的子任务都被选中了
    wrapper.vm.todos.forEach(todo => {
      expect(todo.done).toBeTruthy()
    })

    // 取消完成状态
    await toggleAll.setChecked(false)
    wrapper.vm.todos.forEach(todo => {
      expect(todo.done).toBeFalsy()
    })
  })

  test('Toggle All State', async () => {
    const toggleAll = wrapper.find('input[data-testid="toggle-all"]')
    // 让所有任务都变成完成状态
    wrapper.vm.todos.forEach(todo => {
      todo.done = true
    })
    await Vue.nextTick()
    // 断言 toggleAll 也选中了
    expect(toggleAll.element.checked).toBeTruthy()

    // 取消某个任务未完成，断言 toggleAll 未完成
    wrapper.vm.todos[0].done = false
    await Vue.nextTick()
    expect(toggleAll.element.checked).toBeFalsy()
  })

  test('Clear All Completed', async () => {
    wrapper.vm.handleClearCompleted()
    await Vue.nextTick()
    expect(wrapper.vm.todos).toEqual([
      { id: 1, text: 'eat', done: false },
      { id: 3, text: 'sleep', done: false }
    ])
  })

  test('Filter Todos', async () => {
    // 将路由导航到 /，断言 filterTodos = 完成的任务列表
    wrapper.vm.$route.path = '/'
    await Vue.nextTick()
    expect(wrapper.vm.filterTodos).toEqual([
      { id: 1, text: 'eat', done: false },
      { id: 2, text: 'play', done: true },
      { id: 3, text: 'sleep', done: false }
    ])

    // 将路由导航到 /active
    // 断言 filterTodos = 所有未完成任务
    wrapper.vm.$route.path = '/active'
    await Vue.nextTick()
    expect(wrapper.vm.filterTodos).toEqual([
      { id: 1, text: 'eat', done: false },
      { id: 3, text: 'sleep', done: false }
    ])

    // 将路由导航到 /completed
    // 断言 filterTodos = 所有已完成任务
    wrapper.vm.$route.path = '/completed'
    await Vue.nextTick()
    expect(wrapper.vm.filterTodos).toEqual([
      { id: 2, text: 'play', done: true }
    ])
  })

  test('snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})
