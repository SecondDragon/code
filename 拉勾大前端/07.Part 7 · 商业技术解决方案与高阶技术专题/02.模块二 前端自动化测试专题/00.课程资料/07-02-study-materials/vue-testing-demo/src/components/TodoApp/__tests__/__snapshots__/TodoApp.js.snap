// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoApp.vue snapshot 1`] = `
<section class="todoapp">
  <todoheader-stub></todoheader-stub>
  <section class="main"><input data-testid="toggle-all" id="toggle-all" type="checkbox" class="toggle-all"> <label for="toggle-all">Mark all as complete</label>
    <ul class="todo-list">
      <todoitem-stub todo="[object Object]"></todoitem-stub>
      <todoitem-stub todo="[object Object]"></todoitem-stub>
      <todoitem-stub todo="[object Object]"></todoitem-stub>
    </ul>
  </section>
  <todofooter-stub todos="[object Object],[object Object],[object Object]"></todofooter-stub>
</section>
`;
